core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -114,43 +117,51 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
-                    missingCatalogs.stream()
-                            .map(catalog ->
-                                    Futures.submit(() -> {
-                                        catalogs.computeIfAbsent(catalog.catalogHandle(), ignore -> {
-                                            CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                                            log.debug("Added catalog: " + catalog.catalogHandle());
-                                            return newCatalog;
-                                        });
-                                    }, executor))
-                            .collect(toImmutableList()));
-
-            Deque<Throwable> catalogLoadingExceptions = new LinkedList<>();
-            for (ListenableFuture<Void> loadedCatalog : loadedCatalogs) {
-                try {
-                    loadedCatalog.get();
-                }
-                catch (ExecutionException e) {
-                    if (e.getCause() != null) {
-                        catalogLoadingExceptions.add(e.getCause());
+
+            List<ListenableFuture<CatalogConnector>> loadingFutures = new ArrayList<>();
+            for (CatalogProperties catalogProperties : missingCatalogs) {
+                loadingFutures.add(catalogs.computeIfAbsent(catalogProperties.catalogHandle(), ignored -> loadCatalog(catalogProperties)));
+            }
+
+            CountDownLatch catalogsLoading = new CountDownLatch(loadingFutures.size());
+
+            List<Throwable> catalogLoadingExceptions = new ArrayList<>();
+            for (ListenableFuture<CatalogConnector> catalogFuture : loadingFutures) {
+                addCallback(catalogFuture, new FutureCallback<>()
+                {
+                    @Override
+                    public void onSuccess(CatalogConnector result)
+                    {
+                        catalogsLoading.countDown();
                     }
-                    else {
-                        catalogLoadingExceptions.add(e);
+
+                    @Override
+                    public void onFailure(Throwable throwable)
+                    {
+                        if (throwable.getCause() != null) {
+                            catalogLoadingExceptions.add(throwable.getCause());
+                        }
+                        else {
+                            catalogLoadingExceptions.add(throwable);
+                        }
+
+                        catalogsLoading.countDown();
                     }
-                }
-                catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    catalogLoadingExceptions.add(e);
-                }
-                finally {
-                    loadedCatalog.cancel(true);
-                }
+                }, executor);
+            }
+
+            try {
+                catalogsLoading.await();
```

Comment:
What is the latch for?

the latch has no benefit over doing
```
for f in loadingFutures {
 f.get();
}
```

we also don't need `catalogLoadingExceptions` list.
the first exception received is enough to propagate
after receiving first exception we should cancel all remaining futures 
(and do something to ensure the concurrently loaded catalogs are shut down properly)